src/Compilers/CSharp/Test/Symbol/Symbols/Source/BaseClassTests.cs

```diff
@@ -2408,5 +2408,81 @@ class A : I<System.ValueTuple<object, A.B>>
                 // class A : I<System.ValueTuple<object, A.B>>
                 Diagnostic(ErrorCode.ERR_CircularBase, "B").WithArguments("A", "A").WithLocation(4, 41));
         }
+
+        [Fact]
+        public void TestBadBaseClassVisibility()
+        {
+            var source = @"
+class A { }
+public class B : A { }";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (2,14): error CS0060: Inconsistent accessibility: base class 'A' is less accessible than class 'B'
+                //     public class B : A { }
+                Diagnostic(ErrorCode.ERR_BadVisBaseClass, "B").WithArguments("B", "A").WithLocation(3, 14));
+        }
+
+        [Fact]
+        public void TestBadTypeParameterVisibility()
+        {
+            var source = @"
+class A { }
+public class B<T> { }
+public class C : B<A> { }";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (4,14): error CS9335: Inconsistent accessibility: type 'A' is less accessible than class 'C'
+                //     public class C : B<A> { }
+                Diagnostic(ErrorCode.ERR_BadVisBaseType, "C").WithArguments("C", "A").WithLocation(4, 14));
+        }
+
+        [Fact]
+        public void TestBadContainedTypeVisibility()
+        {
+            var source = @"
+public class A { internal class B { } }
+public class C : A.B { }";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,14): error CS0060: Inconsistent accessibility: base class 'A.B' is less accessible than class 'C'
+                //     public class C : A.B { }
+                Diagnostic(ErrorCode.ERR_BadVisBaseClass, "C").WithArguments("C", "A.B").WithLocation(3, 14));
+        }
+
+        [Fact]
+        public void TestBadContainingTypeVisibility()
+        {
+            var source = @"
+class A { public class B { } }
+public class C : A.B { }";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (3,14): error CS9335: Inconsistent accessibility: type 'A' is less accessible than class 'C'
+                //     public class C : A.B { }
+                Diagnostic(ErrorCode.ERR_BadVisBaseType, "C").WithArguments("C", "A").WithLocation(3, 14));
```

Comment:
I'm wondering whether the message could be improved - since we squiggle the class `C` itself, the user might be confused where to search for the less accessible type (in nested types, in members, in base types, in attributes, ...) - i.e., should the message be something like this?

"Inconsistent accessibility: type A which appears in the base type is less accessible than class C"

Reply:
Both messages work for me, thanks.